### DISCUSSION TITLE
stories_things/cpp-things/sizeof/

### DISCUSSION THREAD

**Discussion by giscus** [Light-City/light-comments#9] [Announcements]
# stories_things/cpp-things/sizeof/

Stories About C Plus Plus

https://light-city.github.io/stories_things/cpp-things/sizeof/

**Comment by cxz0218** [comment]
原则4与5，vs studio运行出来的结果是8 16 16 8，麻烦作者可以修正一下~

**Reply by cxz0218** [reply]
不好意思，是8 16 12 8，麻烦作者可以修正一下~

**Reply by Light-City** [reply]
这个是依据平台来的，x86 64 or 32不太一样

**Reply by Light-City** [reply]
例如：你可以在这里运行：https://wandbox.org/

**Reply by galaxy-git** [reply]
不是平台的问题，是字节对齐的问题，对齐是按2.4.8.16来的，char类型对齐到2字节，所以B的大小是16字节

**Reply by Light-City** [reply]
你确定？那按照你说的对齐两个字节，怎么算出16？ 我的解释是long的大小不一样导致，例如：32位long是4字节.64位是8字节，B按照8字节对齐，char与long要对齐到8，long占4，所以char要对齐到4，最后就是 8➕4➕4等于16，如果是64位操作系统，就是8➕8➕8，因为long是8字节……

**Reply by galaxy-git** [reply]
要确定是多少用gcc编译的时候可以改参数，按照32位和64位编译，到时就看出来是多少了

**Comment by cxz0218** [comment]
我这个是64机器，应该是最常见的最大数值的情况了叭？
以及，感谢回答，这个教程我很喜欢~
&nbsp;


日光沉寂的半海
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2024年3月26日(星期二) 下午2:38
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [Light-City/light-comments] stories_things/basic_content/sizeof/ (Discussion #9)





 
这个是依据平台来的，x86 64 or 32不太一样

**Reply by Light-City** [reply]
你可以看一下sizeof(long)是不是等于4

**Reply by cxz0218** [reply]
sizeof(long)输出的是8，我的ide是visual studio 2022，前面一模一样的代码确实跑出来是8 16 12 8

**Reply by cxz0218** [reply]
https://wandbox.org/ 这里运行确实是你的答案，但是我的8 16 12 8也确实是在64位机器上跑出来的，sizeof(long)是8，这个毫无疑问

**Reply by Light-City** [reply]
那就不科学了，运行sizeof（B）、size of long环境一样不 接图看看

**Comment by jibinghu** [comment]
4.的class C1:public A应修改为class C1:public A1

**Reply by Light-City** [reply]
对，我一会修复，感谢！

**Comment by Austin-Jack** [comment]
原则1的文件名改为blankclass是不是要更贴切些，还有就是为啥要引入bits/stdc++.h呢

**Comment by Austin-Jack** [comment]
对于原则6中的C类，我感觉说明为什么C类本身的vptr不占空间会更好。
这是来自gpt的回答：
> 当一个类派生另一个含有虚函数的类时，它本身的虚函数表指针（vptr）不会占用额外的空间，因为派生类会共享基类的虚函数表。

**Comment by Light-City** [comment]
你确定？那按照你说的对齐两个字节，怎么算出16？
我的解释是操作系统不同，32位long是4字节.64位是8字节，B按照8字节对齐，char与long要对齐到8，long占4，所以char要对齐到4，最后就是 8➕4➕4等于16，如果是64位操作系统，就是8➕8➕8，因为long是8字节……



发自我的iPhone


------------------ Original ------------------
From: galaxy-git ***@***.***&gt;
Date: Tue,Apr 30,2024 4:42 PM
To: Light-City/light-comments ***@***.***&gt;
Cc: Francis ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [Light-City/light-comments] stories_things/basic_content/sizeof/(Discussion #9)





 
不是平台的问题，是字节对齐的问题，对齐是按2.4.8.16来的，char类型对齐到2字节，所以B的大小是16字节

**Comment by AscendLiu** [comment]
原则6：VS2022跑出来的咋是 8 8 32

**Reply by AscendLiu** [reply]
MInGW 确实是 8 8 16

**Comment by xllzs** [comment]
原则6为啥不是24 ，8+8+8 C自己的vptr怎么没了

**Reply by ThinkWithZ** [reply]
有虚函数才有vptr

**Comment by sgtpepperrr** [comment]
我怎么感觉有点乱呢，为什么作者不区分vfptr/vbptr，而都是用vptr？感觉看的有点乱了。

**Reply by ThinkWithZ** [reply]
MSVC才有这个概念

**Reply by sgtpepperrr** [reply]
soga

**Comment by jibinghu** [comment]
已收到，谢谢。

**Comment by jibinghu** [comment]
已收到，谢谢。
